app/models/host/managed.rb

```diff
@@ -452,13 +450,22 @@ def self.import_host_and_facts(hostname, facts, certname = nil, proxy_id = nil)
     # if we were given a certname but found the Host by hostname we should update the certname
     host.certname = certname if certname.present?
 
+    # if proxy authentication is enabled and we have no puppet proxy set, use it.
+    host.puppet_proxy_id ||= proxy_id
+
     # if proxy authentication is enabled and we have no puppet proxy set and the upload came from puppet,
     # use it as puppet proxy.
-    if facts['_type'].blank? || facts['_type'] == 'puppet'
+    if import_type == 'puppet'
       host.puppet_proxy_id ||= proxy_id
     end
```

Comment:
how does it change behavior? if type is puppet then associate a proxy, the only change is that we don't get all facts in this method but we use import_type directly, I don't see why it's not refactoring